Add deprecated-api-versions policy blogpost (#149)
kubewarden/kubewarden.io: content/blog/2022/11/deprecation-policy.md

---
title:  Keeping track of Kubernetes deprecated resources
authors:
- Víctor Cuadrado Juan
date: 2022-11-04
---

It's fact of life: As the Kubernetes API evolves, it's periodically reorganized
or upgraded. This means the old API is [deprecated and later removed](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api).

We deserve to easily keep track of those deprecations and removals. For that, we
have just released
the [deprecated-api-versions policy](https://github.com/kubewarden/deprecated-api-versions-policy/).

### A look at the deprecated-api-versions policy

This policy detects the usage of Kubernetes resources that have been deprecated

> or removed in the API.

Reviewer: ```suggestion
or removed from the Kubernetes API.
```

> It has 2 settings:

Reviewer: ```suggestion
The policy has two settings:
```

> - `kubernetes_version`: The starting version from where to detect deprecated or
>   removed Kubernetes resources.

Reviewer: ```suggestion
- `kubernetes_version`: The starting version from where to detect deprecated or
  removed Kubernetes resources. This setting is mandatory.
```

> - `deny_on_deprecation:` If true, it will deny the operation on the resource
>   (normally its creation). If false, it will just log about it. Defaults to
>   true.

Reviewer: ```suggestion
- `deny_on_deprecation:` If true, it will deny the operation on a resource
  that has been deprecated but not yet removed from the Kubernetes version
  specified by `kubernetes_version`. This setting is optional, it is set to `true` by default.
```

### In action

As usual, instantiate a `ClusterAdmissionPolicy` (cluster-wide) or
`AdmissionPolicy` (namespaced) that makes use of the policy.

For this example, let's work in a k8s cluster of version `1.24.0`.

Here's a definition of a cluster-wide policy that rejects resources that were
deprecated or removed in Kubernetes version `1.23.0` and earlier:

```yaml
kubectl apply -f - <<EOF
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: my-deprecated-api-versions-policy
spec:
  module: ghcr.io/kubewarden/policies/deprecated-api-versions:v0.1.0-k8sv1.26.0
  mutating: false
  rules:
  - apiGroups: ["*"]
    apiVersions: ["*"]
    resources: ["*"]
    operations:
    - CREATE
    - UPDATE
  settings:
    kubernetes_version: "1.23.0"
    deny_on_deprecation: true
EOF
```

> **Info:** In `spec.rules` we are checking every resource in every
> apiGroup and apiVersions. We are doing it for simplicity in this example, yet
> the policy
[metadata.yaml](https://github.com/kubewarden/deprecated-api-versions-policy/blob/b26633515de367cf77b79fb909461a4df6e0e2aa/metadata.yml)
> comes with long and complete, machine-generated `spec.rules` that covers just

> > the resources that are deprecated.

Reviewer: ```suggestion
> the resources that are deprecated.
>
> You can obtain the right rules by using the `kwctl scaffold` command.
```

Author: doh, true, indeed!

The policy has two settings:

- `kubernetes_version`: The starting version from where to detect deprecated or

>   removed Kubernetes resources. This setting is mandatory.

Reviewer 2: does this default to a particular value? If yes, that would be helpful to add.

Author: It doesn't. I suppose we need the user to decide on what version to test against.
